src/osp/string_concat.h

```diff
@@ -0,0 +1,162 @@
+#define INCLUDED_OSP_STRING_CONCAT_H_C0CC2473_002D_406E_99B6_E233565EAA1E
+#pragma once
+
+#include <iterator> // for std::data(), std::size()
+#include <type_traits> // for std::is_pointer<>, std::is_array<>
+
+namespace osp
+{
+
+/**
+ * Gets a pointer to the underlying data of a string type object
+ *
+ * If the parameter is a char *, or a char[], then this
+ * function simply returns the parameter.
+ *
+ * If the parameter is neither a char* or char[], then return
+ * the result of calling the parameters .data() function.
+ *
+ * If the parameter has no .data() function, a compile error results.
+ *
+ * The returned pointer provides no guarantee on the string being nul
+ * terminated. Always use string_data() with string_size() to ensure
+ * no out of bounds data access occurs.
+ */
+template<class STRING_T>
+constexpr decltype(auto) string_data(const STRING_T& str)
+{
+    if constexpr(std::is_pointer_v<STRING_T>)
+    {
+        return str;
+    }
+    else
+    {
+        return std::data(str);
+    }
+} // string_data()
+
+
+//-----------------------------------------------------------------------------
+
+/**
+ * Gets the size of the string provided as the parameter
+ *
+ * If the parameter is a char *, or a char[], then this
+ * function calls the string length function, and therefore
+ * the provided string must be nul-terminated.
+ *
+ * (Future enhancement will allow the size of char[] parameters
+ *  to be statically determined at compile time)
+ *
+ * If the parameter is neither a char* or char[], then return
+ * the result of calling std::size().
+ */
+template<class STRING_T>
+constexpr size_t string_size(const STRING_T & str)
+{
+    if constexpr(std::is_pointer_v<STRING_T>)
+    {
+        return std::char_traits<std::remove_pointer_t<STRING_T>>::length(str);
+    }
+    else if constexpr(std::is_array_v<STRING_T>)
+    {
+        // TODO: Throw if the array contains an embedded nul character.
+        // TODO: Return the static size based on template deduction.
+        return std::char_traits<std::remove_pointer_t<std::decay_t<STRING_T>>>::length(str);
+    }
+    else
+    {
+        return std::size(str);
+    }
+} // string_size()
+
+
+//-----------------------------------------------------------------------------
+
+/**
+ * Effecienctly appends multiple strings together using, at most, a single
+ * allocation to reserve the necessary space, via the "reserve" function call.
+ *
+ * \arg dest   -- The destination where all of the appended strings will be stored
+ * \arg others -- The strings to append into the destination.
+ *
+ * \returns Nothing.
+ */
+template<typename DESTINATION_T, typename ... STRS_T>
+constexpr void string_append(DESTINATION_T & dest, const STRS_T& ... others)
+{
+    // C++17 fold for summation
+    dest.reserve( ( string_size(dest) + ... + string_size(others) ) );
+
+    // C++17 fold for function calls.
+    ( dest.append(string_data(others), string_size(others)), ... );
```

Comment:
I have no idea what this does, can I take your word that it works? If so we should merge this so I can use it in my PR

Reply:
Basically it expands the parameter pack.

Imagine you have others = string1, string2, string3.

If so, then

( dest.append(string_data(others), string_size(others)), ... );

turns into
dest.append(string_data(string1), string_size(string1));
dest.append(string_data(string2), string_size(string2));
dest.append(string_data(string3), string_size(string3));

Reply:
As far as I'm aware, this should work, but obviously if you run into problems with it let me know.

Reply:
@Capital-Asterisk you can merge this whenever you get a chance to look at it